Implement a piecemeal tree file system.

IMarvinTPA opened the pull request:
This allows smaller trees to be integrated into one bigger one.  Good
for distributed tree creation or users who wish to keep a customization
that won't get squashed by an update.

This was inspired by https://github.com/Inventory-Tweaks/inventory-tweaks/issues/355 and a hope that allowing smaller separate files would be easier to foster community building to get the mythical nearly-complete sorting file.

Feedback requested.

Thanks,
IMarv

IMarvinTPA committed:
Implement a piecemeal tree file system.

This allows smaller trees to be integrated into one bigger one.  Good
for distributed tree creation or users who wish to keep a customization
that won't get squashed by an update.

IMarvinTPA committed:
Changed the wrong tree files.

Updated Modded Item Tree file.
Removed unused import.

IMarvinTPA committed:
Exclude tree files because some are too big.

Kobata commented on src/main/java/invtweaks/InvTweaksConfigManager.java:
```diff
@@ -251,5 +303,44 @@ private boolean extractFile(@NotNull ResourceLocation resource, @NotNull File de
         }
     }
+    
+    private List<ResourceLocation> getResourceNames(String inPath) {
```
This seems fairly awkward: It'd allow overriding the included ones with a resource pack (as is the case with the default tree, and why it's loaded that way), but there's no facility in place to mark new files matching from other resource packs.

IIRC the closest thing in vanilla is the language loader and I believe adding new ones requires checking a specific file in the pack that lists the properties of the new language.

Kobata replied:
I suppose I should also point out that feature of loading defaults from resource packs was slightly ill-conceived and mildly pointless, because you tend to load the game without any resource packs set up initially, which means it loads the included defaults and then doesn't bother checking again unless the files are deleted.

IMarvinTPA replied:
I honestly hadn't considered files coming from other resource packs.  I did it so I could get the seed files out of the jar and into place to be used.  What should I consider instead?

Kobata replied:
It's been forever since I've looked at the resource stuff but I don't remember it being simple to enumerate there.

Since the specifics of *why* it's using ResourceLocation aren't really important (I don't think anyone's really using it, and it's awkward to set up a resource pack correctly to work, *and* since it all relies on extracted files it's unlikely to trigger) it's probably safe to leave as-is.

Like I said, it's a weird thing that sounded like a neat idea but in practice it's not really an advantage due to the implementation.

IMarvinTPA replied:
I looked for an enumeration function and there (still) isn't one.

Kobata commented:
The biggest question/concern I've got about this in general is how the merging generates tree-order but that's a thing I'll probably need to actually get things set up to test (or at least try to get a peek at the apparently generated merged tree?)

IMarvinTPA commented:
The result looks a bit like this:
[InvTweaksTree.txt](https://github.com/Inventory-Tweaks/inventory-tweaks/files/2660518/InvTweaksTree.txt)

The process matches categories and appends new nodes as best as it can.  The minecraft.tree file is always the first one loaded and has all of the category nodes, dictionary nodes, class nodes, comment nodes, and id='minecraft:*' nodes.  It has caused some minor reshuffling of items compared to the moddeditemtree.xml file.  I'm looking for the spots where the change is too much and adding categories to help them land closer to where I really want them.  But the actual tee order gets built just like it always has, just with a itemtree file built from parts just before the load happens.

I have built myself a tool in Access that lets me work on the tree visually.  I have added a button that does the saving process that splits the tree by mod.

Kobata commented:
Alright, I've gotten a test setup back up again for all this.

Messing with using partial sets it seems to turn out decently with the files provided: However, it seems like it relies a lot on empty entries in `minecraft.tree` to correctly position sub-categories.

This is the type of thing I was trying to refer to earlier with the ordering: with a fairly easy way of adding new fragments having to update a master file to insure categories are in the right place is still a bit more work than it should be.

This also manifests in there being a bunch of fairly extraneous categories (like all the `swordDamage*` ones) that just seem to exist to try to shove everything into a reasonable order.

IMarvinTPA commented:
The "swordDamage" families are relics from when I was trying to group them manually.  The new sorting doesn't need them as long as they all get grouped in the super-group of sword which has mergeChildren='true' which puts them all in the same order slot, forcing the use of the meleeDamage sort stage.  But that's more an exception to the general rule you are describing.
Perhaps if there were a phase at the end that pruned empty categories before saving?

IMarvinTPA committed:
Add an attribute to help build the order of nodes in the composite tree.

The moddeditemtree.xml file contains the reference tree that I used to
make the individual trees.
The initializer only attempts to extract the tree files if
minecraft.tree is missing and the whole system is only used if InvTweaks
config folder exists.
TreeBuilder updated to read and use treeOrder attribute to fine-tune the
insertion of nodes in the composite tree.

IMarvinTPA commented:
So, I may have gone with a completely different idea.  Each node has a treeOrder attribute, and these are compared to work out where to insert the new node into the big tree.
The Tree loader has been modified to assume anything that is not an ID, OreDict, or Class must be a category since the if condition was just getting complicated for no valuable reason.

IMarvinTPA committed:
Make the treeOrder merge smarter when dealing with missing treeOrders.

Kobata commented:
Alright, after a bunch of thinking there's a few major reservations about all this. (Mostly about the organization of the files)

Ideally a system like this would:
- Use the current vanilla tree as the major base
- Not require any coordination between separate files as much as possible (e.g., you can't really help requiring fairly matching categories but things like your order attribute are a bit much)
- Probably undergo maintenance/development of the additional files somewhat separately (release cadence will probably be different enough that it's not really worth trying to embed all of them in the mod jar), if there's really a major desire for a 'blessed' version I can create a new repo under the organization for this purpose.

It also seems rather pointless to build up a separate 'merged' file when it's essentially being rebuilt every time anyway, as there doesn't seem to be any useful saved information to prevent rebuilding it when no files are changed so it has to assume there's been a change.

If everything ends up working decently then the loading from a single tree file can be essentially depreciated with some migration code to move it to `minecraft.tree` and additional files can be obtained by users or modpack authors to add on to that from a source that is likely to be more up to date than whatever happened to be current upon the production of a release.

IMarvinTPA commented:
It only re-builds it the first time on load, after that it only does it when it detects a change in one of the files.  The only reason it does it the first time, per load, is that I can't tell if a new file was added or any other files were changed while off-line.  The merged file is the one actually being read by the tree loader since it has everything in the correct order.

I'm not sure I've ever seen any modpacks with customized trees.  I have an unreleased Netherdeep update that was going to have one.

I'm thinking about dropping all of my tree files, promoting the baseline xml file as minecraft.xml, moving all of the InvTweaks files to the InvTweaks config folder, adding a sub folder called trees for the tree files.  Making the "Get a modded tree file" button print a link to a new page on the wiki that can have links to attempts to build better trees the user can download and put into the tree folder.  And add instructions on how to use the merge system.

Kobata commented:
> It only re-builds it the first time on load

Yes, this is basically what I was talking about: If it's not being used between sessions it's probably better off in-memory only or just an actual temp file. Would help protect people from thinking they can just edit it and get changes overwrite.

> I'm not sure I've ever seen any modpacks with customized trees

Probably not, but since the point of this type of thing is to make it easier to grab things piecemeal if there's an easy source of trees that cover certain mods I would expect more modpacks to do include things.

> Making the "Get a modded tree file" button print a link to a new page on the wiki that can have links to attempts to build better trees the user can download and put into the tree folder.

I was something thinking of something around there, making that either link to a source or if there's some 'blessed' source (probably with things organized such that modids can be used to find files) possibly even having an option to automatically download & load the set for currently installed mods.

> And add instructions on how to use the merge system.

[Inventory-Tweaks/Docs](https://github.com/Inventory-Tweaks/Docs) does probably need an update once everything is settled. If for nothing else with the file names changing.

IMarvinTPA commented:
Repository for my trees: https://github.com/IMarvinTPA/InventoryTweaksTrees

IMarvinTPA committed:
Remove modded tree files.

Make the modded tree button go to a repository of modded trees.
Move the merged tree into the system temp folder to minimize the chance
the user would think they could modify it.
Add a readme to the tree folder.
Move all config files to the InvTweaks config folder.
Make a trees folder in the InvTweaks config folder for the tree files.

IMarvinTPA committed:
Removed unused imports.

IMarvinTPA committed:
Remove unused changes to build.gradle.

IMarvinTPA commented:
I think I'm done.  Let me know if you want me to change anything.  I am keeping the mergeOrder, but it does not appear in the default tree file.

Kobata commented:
Alright, thanks for putting up with me being a bit picky.

IMarvinTPA commented:
Cool.  A level of pickiness is required for quality.

Kobata commented:
I will keep the offer open for setting up a repo in the organization for the modded files if you want it, btw.
